### PR TITLE
[feature]Allow for mixed projects and registrations in a collection [OSF-6029]

### PR DESCRIPTION
**Pull request by TomBaxter**
## Purpose:
[OSF-6029](https://openscience.atlassian.net/browse/OSF-6029)
[My Projects] Registrations that aren't yours have blank contributor names and throw 404 error when expanded

## Changes:
Update  api/base/serializers.py
    Add class LinkedRegistration
    Update class LinkedNodesRelationshipSerializer do not return registrations
    Add class LinkedRegistrationsRelationshipSerializer

Update api/base/views.py
    Update class LinkedNodesRelationship do not return registrations
    Add class LinkedRegistrationsRelationship
    Add class BaseLinkedList

Update api/collections/serializers.py
    Update class CollectionSerializer
        Add attribute linked_registrations
        Update def get_node_links_count do not return registrations
        Add def get_registration_links_count

Update api/collections/urls.py
    Add /linked_registrations/
    Add /relationships/linked_registrations/

Update api/collections/views.py
    update class LinkedNodesList do not return registrations
    add class LinkedRegistrationsList
    add class CollectionLinkedRegistrationsRelationship

Update api/nodes/views.py
    Update class NodeLinkedNodesRelationship
        Update def get_queryset do not return registrations

Update api/registrations/views.py
    Add class LinkedRegistrationsList

Update website/project/model.py
    Update class Node
        Add def linked_registrations_self_url
        Add def linked_registrations_related_url

Update website/static/js/myProjects.js
    Update function NodeFetcher allow for fetching projects and registrations in one fetcher
    Update function buildCollectionNodeData now requires linke_type argument
    Update var MyProjects now handles combinations of projects and registrations
    Update var Collections now passes type to buildCollectionNodeData

Update website/static/js/project-organizer.js
    function _poTitleColumn  now returns type info, for website/static/js/myProjects.js var Collections

Update api_tests/collections/test_views.py
  Added many tests for linked registrations

## Side effects
The following no longer return registrations
LinkedNodesRelationshipSerializer
LinkedNodesRelationship
CollectionSerializer get_node_links_count
LinkedNodesList
NodeLinkedNodesRelationship get_queryset

[#OSF-6029]

### CODE REVIEW

File: api/collections/views.py [hunk 1]
```diff
@@ -378,6 +382,107 @@ def get_parser_context(self, http_request):
...
+    """
+    permission_classes = (
+        drf_permissions.IsAuthenticatedOrReadOnly,
+        ContributorOrPublic,
+        ReadOnlyIfRegistration,
+        base_permissions.TokenHasScope,
+    )
+
+    required_read_scopes = [CoreScopes.NODE_LINKS_READ]
+    required_write_scopes = [CoreScopes.NODE_LINKS_WRITE]
```
Comment (brianjgeiger) [inline api/collections/views.py]: If this is just a listview, the write scope should be `CoreScopes.NULL`.

File: api/registrations/views.py [hunk 2]
```diff
@@ -860,6 +861,54 @@ def get_object(self):
...
+    """
+    permission_classes = (
+        drf_permissions.IsAuthenticatedOrReadOnly,
+        ContributorOrPublic,
+        ReadOnlyIfRegistration,
+        base_permissions.TokenHasScope,
+    )
+
+    required_read_scopes = [CoreScopes.NODE_LINKS_READ]
+    required_write_scopes = [CoreScopes.NODE_LINKS_WRITE]
```
Comment (brianjgeiger) [inline api/registrations/views.py]: `CoreScopes.NULL`

File: api/registrations/views.py [hunk 3]
```diff
@@ -860,6 +861,54 @@ def get_object(self):
         return obj
 
 
+class LinkedRegistrationsList(JSONAPIBaseView, generics.ListAPIView, RegistrationMixin):
```
Comment (brianjgeiger) [inline api/registrations/views.py]: This looks an awful lot like the LinkedRegistrationsList in Collections. Can you DRY this up by making a base version without the RegistrationMixin nor CollectionMixin then just subclass here?

File: api_tests/collections/test_views.py [hunk 4]
```diff
@@ -704,12 +709,15 @@ def test_do_not_return_node_pointers_logged_out(self):
...
         res = self.app.get(self.url, auth=self.user_one.auth)
         res_json = res.json['data']
-        assert_equal(len(res_json), 2)
+        assert_equal(len(res_json), 4)
         assert_equal(res.status_code, 200)
         assert_equal(res.content_type, 'application/vnd.api+json')
         first_embedded = res_json[0]['embeds']['target_node']['data']['id']
         second_embedded = res_json[1]['embeds']['target_node']['data']['id']
-        assert_items_equal([first_embedded, second_embedded], [self.project._id, self.public_project._id])
+        # node_links end point does not handle registrations correctly
```
Comment (brianjgeiger) [inline api_tests/collections/test_views.py]: This should be documented on the relevant view endpoint documentation and we should say that using node_links is deprecated in favor of linked_nodes.

Comment (brianjgeiger) [inline api_tests/collections/test_views.py]: And, of course, linked_registrations.

Comment (brianjgeiger) [inline api_tests/collections/test_views.py]: This might get done in a future ticket.

**Comment by brianjgeiger** [comment]
Overall, I think this looks good. Fix the corescopes for the write scope and dry up the function, then put it in Product Demo and arrange a time with Sara. Pass Done. Don't worry about updating the info for deprecating node_links; I'll arrange a separate ticket for that. 🐧

File: api/collections/views.py (hunk 1 above)
Comment (TomBaxter) [inline api/collections/views.py]: Done

File: api/registrations/views.py (hunk 2 above)
Comment (TomBaxter) [inline api/registrations/views.py]: Done

File: api/registrations/views.py (hunk 3 above)
Comment (TomBaxter) [inline api/registrations/views.py]: Done. Both instances of LinkedRegistrationsList and both instances of LinkedNodesList now subclass BasLinkedList.

File: api_tests/collections/test_views.py (hunk 4 above)
Comment (TomBaxter) [inline api_tests/collections/test_views.py]: Done

File: api/collections/views.py [hunk 5]
```diff
@@ -378,9 +365,101 @@ def get_parser_context(self, http_request):
...
+        res['is_relationship'] = True
+        return res
+
+
 class NodeLinksList(JSONAPIBaseView, bulk_views.BulkDestroyJSONAPIView, bulk_views.ListBulkCreateJSONAPIView, CollectionMixin):
     """Node Links to other nodes. *Writeable*.
 
+    # Deprecated
+
+    The use of /collections/ID/node_links/ is deprecated in favor of linked_nodes, linked_registrations or a combination of both.
```
Comment (brianjgeiger) [inline api/collections/views.py]: 👍

File: api_tests/collections/test_views.py (hunk 4 above)
Comment (brianjgeiger) [inline api_tests/collections/test_views.py]: 👍

**Comment by TomBaxter** [comment]
There will be merge conflicts between this PR and [PR6308](https://github.com/CenterForOpenScience/osf.io/pull/6300)

**Pull request merged**